stores/sql/sqlite/main.go

```diff
@@ -701,6 +658,70 @@ func (tx *MainDatabaseTx) RemoveOfflineHosts(ctx context.Context, minRecentFailu
 	return ssql.RemoveOfflineHosts(ctx, tx, minRecentFailures, maxDownTime)
 }
 
+// RenameDirectories renames all directories in the database with the
+// given prefix to the new prefix.
+func (tx *MainDatabaseTx) RenameDirectories(ctx context.Context, bucket, prefixOld, prefixNew string) (int64, error) {
+	// prepare statements
+	updateNameStmt, err := tx.Prepare(ctx, "UPDATE directories SET name = ? WHERE id = ?")
+	if err != nil {
+		return 0, err
+	}
+	defer updateNameStmt.Close()
+
+	updateParentStmt, err := tx.Prepare(ctx, "UPDATE directories SET db_parent_id = ? WHERE db_parent_id = ?")
+	if err != nil {
+		return 0, err
+	}
+	defer updateParentStmt.Close()
+
+	existsStmt, err := tx.Prepare(ctx, "SELECT id FROM directories WHERE name = ? AND db_bucket_id = (SELECT id FROM buckets WHERE buckets.name = ?)")
+	if err != nil {
+		return 0, err
+	}
+	defer existsStmt.Close()
+
+	// fetch directories
+	rows, err := tx.Query(ctx, "SELECT id, name FROM directories WHERE name LIKE ? AND db_bucket_id = (SELECT id FROM buckets WHERE buckets.name = ?)", prefixOld+"%", bucket)
+	if err != nil {
+		return 0, err
+	}
+	defer rows.Close()
+
+	// rename directories
+	for rows.Next() {
```

Comment:
This is another spot that might be causing issues by not consuming a cursor before querying and updating within the same txn. 

I'm also surprised that this method doesn't update any objects. I'd imagine that if you try to update a src directory to a dst dir that already exists, you'd need to:

1. update the children of the src dir to point to the dst dir (`updateParentStmt`)
2. update all objects that are in the src directory to the dst one (missing?)
3. delete the src dir (missing?)

Not sure why especially 2. doesn't cause issues though.

Reply:
I'll check out the cursor thing. With regards to your other remarks, `RenameObjects` looks like this:
```go
return s.db.Transaction(ctx, func(tx sql.DatabaseTx) error {
		dirID, err := tx.RenameDirectories(ctx, bucket, prefixOld, prefixNew)
		if err != nil {
			return err
		} else if err := tx.RenameObjects(ctx, bucket, prefixOld, prefixNew, dirID, force); err != nil {
			return err
		}
		s.triggerSlabPruning()
		return nil
	})
```

it's a 2-step process within a single transaction. The reason is two-fold, it's a lot of code so I just naturally defined a `RenameDirectories`, but also `RenameDirectories` can be in the common sql while `RenameObjects` has a MySQL and SQLite version. So it's actually deduplicating code this way. I don't remove directories since we prune them anyway and they're not doing any harm there.